src/turbomind/models/llama/SequenceManager.cc

```diff
@@ -401,14 +401,26 @@ auto SequenceManager::Materialize(Sequences                    sequences,
 
     if (block_trie_->enabled()) {
         // verify blocks in trie cache
-        block_trie_->verify();
+        int valid_count = block_trie_->verify();
+        if (rank_ == 0) {
+            TM_LOG_INFO("[PrefixCache] #all_cached_blocks: %d", valid_count - 1);
+        }
 
         // match prefix cache
         for (int i = 0; i < sequences.size(); i++) {
             if (!sequences[i]->prompt.empty() && sequences[i]->blocks.empty()) {
                 auto& seq = const_cast<Sequence&>(*sequences[i]);
                 block_trie_->match(seq);
                 seq.cache_len = seq.blocks.size() * block_seq_len_;
+                if (rank_ == 0) {
```

Comment:
Printing for each sequence, will it be costly? May we consider doing a complete count and then printing only once? What is your suggestion?

Reply:
when `tp>1`, for each device, the count will be the same, so we don't need to print all of them

Reply:
My point is that the statistics may be calculated and printed once for each batch, not for each sequence and has nothing to do with tp.

Reply:
Do you mean we can calculate and log an average hit_rate for a batch of sequences in one iteration?

Reply:
I think this may be acceptable. The amount of logs has decreased significantly, and at the same time, we can also intuitively understand the hit rate at the batch level. Do you have any suggestions? @lzhangzz @irexyc @lvhan028

Reply:
cc @josephrocca

Reply:
@zhyncs For my use case, what would be the most useful is to simply add the number of prefix tokens that were cached to the API response. E.g. the last EventStream object. Otherwise I need need to write my own parser for the logs to simply get the info on each request.

E.g. something like this would be great as the *final* EventStream item:
```
{
  "id": "15964",
  "object": "text_completion",
  "created": 1720934754,
  "model": "repo/modelname", 
  "choices": [{"index":0,"text":"\n","logprobs":null,"finish_reason":"stop"}],
  "request_summary": {
    "input_tokens": 2542,
    "output_tokens": 230,
    "prefix_cached_tokens": 1740,
    "prefill_time": 0.4,
    "decoding_time": 2.3
  }
}
```
I've added some other example info there that would be useful to me, just as an example of how `request_summary` (or similar) could be extended in the future - a couple inspired by this issue:

* https://github.com/vllm-project/vllm/issues/4683

There is precedent for adding extra info to the final EventStream object - e.g. `generated_text` in final TGI EventStream, and `usage` in final OpenAI API EventStream:

> An optional field that will only be present when you set `stream_options: {"include_usage": true}` in your request. When present, it contains a null value except for the last chunk which contains the token usage statistics for the entire request.

This reason this would be beneficial for me is because I collect statistics at my load balancing server, keyed by GPU type, hosting providers, etc. - and if I simply needed to inspect the `request_summary`  field of the final EventStream object, then this makes things very easy for me.

However, if this is not possible, I will certainly benefit from the current logging approach. Thanks for your work on this @ispobock!

Reply:
Hi @josephrocca Adding `prefix_cached_tokens` to the `usage` of the final EventStream item is technically feasible. But this will involve API changes. Currently, only non-streaming responses have `usage` field, which contains request-level information such as `prompt_tokens`, `completion_tokens`, and `total_tokens`. @AllentDan  may help confirm this. Whether to add `prefix_cached_tokens` to `usage` and whether to attach `usage` to the streaming response still need to be discussed.
cc: @lvhan028 @lzhangzz @irexyc @zhyncs

Reply:
@ispobock I think the streaming response returns the "usage" when the last token is generated.